packages/qwik-city/src/runtime/src/server-functions.ts

```diff
@@ -224,53 +229,73 @@ export const validatorQrl = ((
 /** @public */
 export const validator$: ValidatorConstructor = /*#__PURE__*/ implicit$FirstArg(validatorQrl);
 
+const flattenValibotIssues = (issues: v.GenericIssue[]) => {
+  return issues.reduce<Record<string, string | string[]>>((acc, issue) => {
+    if (issue.path) {
+      const hasArrayType = issue.path.some((path) => path.type === 'array');
+      if (hasArrayType) {
+        const keySuffix = issue.expected === 'Array' ? '[]' : '';
+        const key =
+          issue.path
+            .map((item) => (item.type === 'array' ? '*' : item.key))
+            .join('.')
+            .replace(/\.\*/g, '[]') + keySuffix;
+        acc[key] = acc[key] || [];
+        if (Array.isArray(acc[key])) {
```

Comment:
am I reading this correctly that it won't tell you which array item has the issue? Is that not possible to do?

Reply:
That's true and I think it's bad, but I just followed the Zod implementation because changing it would lead to breaking changes due to the `ValidatorErrorKeyDotNotation` type.

Reply:
Hmm, could we not have a different error signature for valibot? Make it a generic type?

Reply:
We could do this, but it would lead to a major rewrite of the types in `types.ts`. Also, switching from `zod$` to `valibot$` gets confusing as the error implementation is inconsistent. I have heard that there are plans to rewrite Qwik City. This is something we could tackle as part of that.

Reply:
I recommend choosing a unified format without edge cases, or making it custom and just building on top of the schema library's error formatting. Both Zod and Valibot contain a `flatten` function/method.

Reply:
I prefer a per-validator format. That way if people want to roll their own they can, and valibot can already provide nicer errors.

Reply:
I agree, but how should I proceed with this PR? I don't think we should break the current `zod$` implementation. Also, it might be complicated to change this only for Valibot because we probably need to change/break a few types.

Reply:
But if each validator has its own return type, we're not breaking zod$ are we?

Reply:
The current `zod$` implementation used generic names for types like `GetValidatorType` and `TypedDataValidator`. We could of course add `GetValibotValidatorType` and `TypedValibotDataValidator` and add new overload signatures only for Valibot, but that would make the code even more complicated.

Since `zod$` does not use a Zod-specific return type, I just "plugged" the `valibot$` implementation into the `zod$` one and improved a few things. It might be easier to stick with my current solution for now and rewrite it at some point.

Reply:
But if we do it now, it's a new feature, and if we do it later, it's a breaking feature, no?

Reply:
Yes, that's true, but we'll probably break it for `zod$` at some point anyway.

Reply:
If we have this generic per-validator API we wouldn't have to, we'd just make zod2$.

Reply:
That's true, but I'm not sure we should increase the number for each breaking change. In the end, I'm fine with either solution. So feel free to make a final decision, and I will try my best to implement it.

Reply:
Another breaking change we should probably consider in the long run is to remove the export of Zod's `z`. I am not sure if such a library should be "baked" into Qwik City. It's probably better to just use the version of Zod that the user has deliberately installed. But that's another topic.

Reply:
If we want to strictly follow semver we can create zod2$ and in V2 deprecate zod2$ and use zod2$'s code for zod$. That way we don't have to do a breaking change for valibot$ and it can start to share code with zod2$ (+ the benefits for users of zod2$) right now.

But if it's just easier for now to go with zod$ and valibot$ that's good too. Valibot users are quite bleeding-edgy so they shouldn't complain if there's a breaking change for valibot$ in the future.

I think you can take the decision @fabian-hiller since you have more context. Both are fine imo. Easier and cleaner code might be more important than another optimization atm.

> Another breaking change we should probably consider in the long run is to remove the export of Zod's z
Would be good to have those zod breaking changes in the same release. Could you create an issue so that we don't forget?

Reply:
> But if it's just easier for now to go with zod$ and valibot$ that's good too. Valibot users are quite bleeding-edgy so they shouldn't complain if there's a breaking change for valibot$ in the future.

Let’s not adopt a  breaking change is okay mentality from the get-go. The recent Valibot API change was a not-so-small problem in our apps at work recently.